File: (9) Restore IP Addresses/restoreIp.java
import java.util.ArrayList;
import java.util.List;

public class Main {

    public static boolean isValid(String s) {
        if (s.length() > 3 || s.isEmpty() || (s.charAt(0) == '0' && s.length() > 1) || Integer.parseInt(s) > 255) {
            return false;
        }
        return true;
    }

    public static List<String> restoreIpAddresses(String s) {
        List<String> result = new ArrayList<>();
        int n = s.length();

        for (int i = 1; i < 4 && i < n - 2; ++i) {
            for (int j = i + 1; j < i + 4 && j < n - 1; ++j) {
                for (int k = j + 1; k < j + 4 && k < n; ++k) {
                    String a = s.substring(0, i);
                    String b = s.substring(i, j);
                    String c = s.substring(j, k);
                    String d = s.substring(k);

                    if (isValid(a) && isValid(b) && isValid(c) && isValid(d)) {
                        result.add(a + "." + b + "." + c + "." + d);
                    }
                }
            }
        }

        return result;
    }

    public static void main(String[] args) {
        String s = "25525511135"; // Example input
        List<String> ips = restoreIpAddresses(s);

        System.out.print("Valid IP addresses: ");
        for (String ip : ips) {
            System.out.print(ip + " ");
        }
        System.out.println();
    }
}

File: (9) Restore IP Addresses/restoreIp.py
def is_valid(s):
    return len(s) <= 3 and (len(s) == 1 or s[0] != '0') and int(s) <= 255

def restore_ip_addresses(s):
    result = []
    n = len(s)

    for i in range(1, min(4, n - 2)):
        for j in range(i + 1, min(i + 4, n - 1)):
            for k in range(j + 1, min(j + 4, n)):
                a = s[:i]
                b = s[i:j]
                c = s[j:k]
                d = s[k:]

                if is_valid(a) and is_valid(b) and is_valid(c) and is_valid(d):
                    result.append(a + "." + b + "." + c + "." + d)

    return result

if __name__ == "__main__":
    s = "25525511135"  # Example input
    ips = restore_ip_addresses(s)

    print("Valid IP addresses:", ips)
